### PR TITLE
Implemented personality and personality_description parameters

### PR DESCRIPTION
**Pull request by arneboe**
This is my first attempt at personality and personality_description getters/setters.
It is implemented on top of PR https://github.com/someweisguy/esp_dmx/pull/90.
I tested it using my rdm tester and it works.

As with the other PR I am not sure if this is the correct way to implement it. 
I would appreciate it if you review the code and point out anything odd :-)

The PR is kinda hard to read because it contains all the changes from https://github.com/someweisguy/esp_dmx/pull/90.
Reading the diff against https://github.com/someweisguy/esp_dmx/pull/90 is much easier.
(or just read the last commit, same thing :D)

### CODE REVIEW

**Comment by someweisguy** [comment]
Thanks for the hard work on this and #93. I haven't looked closely at these, but if they are working for you then it is likely that this PR is at least 80% complete.

There are a few minor changes that will need to be made but I will share those with you when I can review this PR more closely! I think it would be helpful if I documented some core concepts of the library better so I think I'll add that to my to-do list!

File: src/rdm/responder.c [hunk 1]
```diff
@@ -182,6 +182,144 @@ static int rdm_simple_response_cb(dmx_port_t dmx_num,
...
+{
+  // Return early if the sub-device is out of range
+  if (header->sub_device != RDM_SUB_DEVICE_ROOT) {
+    *pdl_out = rdm_pd_emplace_word(pd, RDM_NR_SUB_DEVICE_OUT_OF_RANGE);
+    return RDM_RESPONSE_TYPE_NACK_REASON;
+  }
+
+  if (header->cc != RDM_CC_GET_COMMAND) {
+    //the supported params list is read-only
+      ESP_LOGE(TAG, "RDM_PID_SUPPORTED_PARAMETERS is read-only");
```
Comment (someweisguy) [inline src/rdm/responder.c]: I don't think this log message is necessary. The RDM controller is the one generating the bad RDM request, not the user. So we shouldn't need to alert the user.

File: src/dmx/hal.c [hunk 2]
```diff
@@ -451,17 +451,20 @@ bool dmx_driver_install(dmx_port_t dmx_num, const dmx_config_t *config,
...
     rdm_register_disc_unique_branch(dmx_num, NULL, NULL);
     rdm_register_disc_un_mute(dmx_num, NULL, NULL);
     rdm_register_disc_mute(dmx_num, NULL, NULL);
     rdm_register_device_info(dmx_num, &device_info, NULL, NULL);
     rdm_register_software_version_label(dmx_num, config->software_version_label,
                                         NULL, NULL);
     rdm_register_identify_device(dmx_num, rdm_default_identify_cb, NULL);
+    rdm_register_dmx_personality(dmx_num, NULL, NULL);
+    rdm_register_dmx_personality_description(dmx_num, NULL, NULL);
+
```
Comment (someweisguy) [inline src/dmx/hal.c:464]: This section should be left to registering the 9 required PIDs. The reason is because if we register all the RDM parameters in `dmx_driver_install()` we may run out of space in the driver.

The intention was for users themselves to call the `rdm_register_` functions for parameters that they wish to register. In other words, if users want the RDM responder to handle `RDM_PID_DMX_PERSONALITY` requests, they will need to call `rdm_register_dmx_personality()` in their own code.

**Review by someweisguy — CHANGES_REQUESTED** [review]
Great work again! Thanks for bearing with my comments. :D

File: src/dmx/hal.c (hunk 2 above)
Comment (arneboe) [reply src/dmx/hal.c:464]: From a user perspective it feels strange to have to provide the parameters myself.
I have to provide all personalities to `dmx_driver_install()`. If the driver asks for the personalities at construction time, I expect that it creates all the necessary parameters to make them available via rdm.

But that is just my opinion. In the end either way is fine with me :) I'll change it. Just wanted to point out that semantically it does make a lot of sense to register them in `dmx_driver_install()`.

The memory argument is kinda invalid because I have to provide all personalities at construction time, thus the strings exist anyway. 
Edit: A well, the entry in the callback list takes up a few bytes so I guess the argument is kinda valid :D

Edit2:
From a rdm-controller point of view it is also a little inconsistent that the device provides the number of personalities and the current personality as part of the device_info by default but does not provide the means to get the personality description and change the personality.

I guess what I am saying is: If the user provides more than one personality, the parameters have to be there anyway because it doesn't make sense to provide several personalities but no way to change the personality.

File: src/rdm/responder.c (hunk 1 above)
Comment (arneboe) [reply src/rdm/responder.c]: agreed

File: src/dmx/hal.c (hunk 2 above)
Comment (someweisguy) [reply src/dmx/hal.c:464]: These are excellent points. I've been thinking about this today and I think your plan makes much more sense than mine. Especially after reading your comments. I apologize!

Do you think it makes more sense to leave these here, or put them in the `if` clause below these lines? In other words, should these responses be registered if the ESP32 does not have a DMX start address?

Per your plan, I think there are a number of PIDs that could be added for newbies to take advantage of, including the PIDs you have added in this and your other PRs. But I am also trying to strike a balance in this library between "easy for newbies" and "powerful for advanced users." I think a great way to achieve this balance would be to add an option in the sdkconfig which could use macros to conditionally compile the registration of these default PIDs. The default value would mean that these PIDs are pre-registered. But if you had an unusual use-case, you couuld turn it off. Thoughts? I am happy to accept this PR as is and implement the sdkconfig feature myself.

Edit: Here's a list of potential pre-registered PIDs. These should not be implemented in this PR but at a later date instead :)

- `RDM_PID_COMMS_STATUS`
- `RDM_PID_QUEUED_MESSAGE`
- `RDM_PID_STATUS_MESSAGE`
- `RDM_PID_STATUS_ID_DESCRIPTION` (maybe?)
- `RDM_PID_CLEAR_STATUS_ID` (if the above is included?)
- `RDM_PID_DEVICE_LABEL`
- `RDM_PID_BOOT_SOFTWARE_VERSION_ID`
- `RDM_PID_BOOT_SOFTWARE_VERSION_LABEL`
- `RDM_PID_BOOT_POWER_CYCLES` (maybe?)

And I'm open to discussing others as well!

**Comment by someweisguy** [comment]
I was commenting about getters/setters in #93 and it made me realize I forgot to address getters/setters in this PR. 

For `RDM_PID_DMX_START_ADDRESS`, there are two getters and two setters. The getters are `dmx_get_start_address()` and `rdm_get_dmx_start_address()`. And there is a similar prefix change for the setter: `dmx_set_start_address()` and `rdm_set_dmx_start_address()`. I implemented `dmx_get_start_address()` and its counterpart, and made `rdm_get_dmx_start_address()` a wrapper function. I also made a wrapper for the setter.

Because DMX start address can be used in a DMX context as well as an RDM context I thought it would be confusing for the user to have to remember that only one set of getter/setter functions existed.

I don't _think_ this is an anti-pattern, but I would love some input if I am mistaken. Should RDM getter/setter wrapper functions also be implemented for the PIDs in this PR? 

Edit: [Here's an example wrapper function](https://github.com/someweisguy/esp_dmx/blob/release/v3.1/src/rdm/parameters.c#L48C11-L57).

File: src/dmx/hal.c (hunk 2 above)
Comment (arneboe) [reply src/dmx/hal.c:464]: > These are excellent points. I've been thinking about this today and I think your plan makes much more sense than mine. Especially after reading your comments. I apologize!

No apologies needed! I really enjoy your detailed answers.

> Do you think it makes more sense to leave these here, or put them in the `if` clause below these lines? In other words, should these responses be registered if the ESP32 does not have a DMX start address?

What is the meaning of not having a start address? The only use-case that I can think of is that the user wants to implement a dmx controller instead of a client. In that case none of the rdm parameters should be registered because there is no device on the bus that could query them.
However, what happens if the user later on sets a dmx address? Does that make any sense?


> Per your plan, I think there are a number of PIDs that could be added for newbies to take advantage of, including the PIDs you have added in this and your other PRs. But I am also trying to strike a balance in this library between "easy for newbies" and "powerful for advanced users." I think a great way to achieve this balance would be to add an option in the sdkconfig which could use macros to conditionally compile the registration of these default PIDs. The default value would mean that these PIDs are pre-registered. But if you had an unusual use-case, you couuld turn it off. Thoughts? I am happy to accept this PR as is and implement the sdkconfig feature myself.

I like it. That way it is easy for the default use-case but experts can still turn it off.
I am traveling for the next two weeks, thus I would be happy if you implemented it because I won't implement anything within the next two weeks :)


 
>     * `RDM_PID_COMMS_STATUS`
Seems like a use-case that is only important for device developers. Since this library is aimed at devs it might make sense. But the data is very specific. I am not sure that it would help me debug any real-world problem.

>     * `RDM_PID_QUEUED_MESSAGE`
Yes, at some point queued messages need to be supported.

 >     * `RDM_PID_STATUS_MESSAGE`
>     * `RDM_PID_STATUS_ID_DESCRIPTION` (maybe?)
>     * `RDM_PID_CLEAR_STATUS_ID` (if the above is included?)
Imo this should not be default. It heavily depends on the device that the user is developing. I.e. error/warning messages will be custom for every device. 

>     * `RDM_PID_DEVICE_LABEL`
This is a must imo because tools/consoles ask for it by default and display the label. Without this the device is only identified using its id which is impossible to remember.

>     * `RDM_PID_BOOT_SOFTWARE_VERSION_ID`
>     * `RDM_PID_BOOT_SOFTWARE_VERSION_LABEL`
Yes, since they are already part of the device_info.
>     * `RDM_PID_BOOT_POWER_CYCLES` (maybe?)
Could be interesting for devices that have some kind of wear and tear. But I cannot imagine a real use-case right now.

**Comment by arneboe** [comment]
> For `RDM_PID_DMX_START_ADDRESS`, there are two getters and two setters. The getters are `dmx_get_start_address()` and `rdm_get_dmx_start_address()`. And there is a similar prefix change for the setter: `dmx_set_start_address()` and `rdm_set_dmx_start_address()`. I implemented `dmx_get_start_address()` and its counterpart, and made `rdm_get_dmx_start_address()` a wrapper function. I also made a wrapper for the setter.
>
> Because DMX start address can be used in a DMX context as well as an RDM context I thought it would be confusing for the user to have to remember that only one set of getter/setter functions existed.
 
 Having two methods that do the same thing can also be confusing. In the end either way is fine as long as it is consistent imo. 
 
 > I don't _think_ this is an anti-pattern, but I would love some input if I am mistaken. Should RDM getter/setter wrapper functions also be implemented for the PIDs in this PR?
 
Since the personality is useful in both contexts, yes.

File: src/dmx/hal.c (hunk 2 above)
Comment (someweisguy) [reply src/dmx/hal.c:464]: >What is the meaning of not having a start address?

I am not sure, to be honest. The RDM specification allows devices to not possess a DMX start address, but I am not sure what it really means. 😆 

My best guess is that it can be useful for some sub-devices or if you are creating an DMX/RDM splitter. Or indeed if you are making a RDM controller, like you mention.

>However, what happens if the user later on sets a dmx address? Does that make any sense?

If a RDM responder does not have a DMX start address, it will reject any requests to change its DMX start address. I tried to make sure the behavior was the same if the user tried to set the DMX start address of the device in code. I _think_ this is what the RDM spec says. i.e., the user can set a DMX start address using `dmx_config_t` in `dmx_driver_install()`. If the user sets a valid DMX start address, calls to `dmx_set_start_address()` will succeed if a valid DMX start address is given. If a valid DMX start address isn't set in `dmx_driver_install()`, calls to `dmx_set_start_address()` will fail.

Comment (someweisguy) [reply src/dmx/hal.c:464]: - `RDM_PID_COMMS_STATUS` I might do this one for my own sake. Debugging can be tedious at time. :D Plus this seems to be another edge-case PID that would require additions to the DMX driver.
 - `RDM_PID_STATUS_MESSAGE` I had the opportunity to test the RDM responder using an [ETC Colorsource](https://www.etcconnect.com/Products/Consoles/ColorSource/Features.aspx) not long ago. It appeared that ETC's RDM discovery algorithm involved the use of this PID. After discovery, the ETC Colorsource refused to send RDM requests to the ESP32. I believe this may have been because the ESP32 did not support `RDM_PID_STATUS_MESSAGE` but I regret that I was unable to do further testing. It also appears that this PID is closely related to queued messages so I may add this when I add queued messages.
- `RDM_PID_STATUS_ID_DESCRIPTION`, `RDM_PID_CLEAR_STATUS_ID` after rereading the spec, fair point on these.
- `RDM_PID_BOOT_SOFTWARE_VERSION_ID` and `RDM_PID_BOOT_SOFTWARE_VERSION_LABEL` are not part of device info. Notice that these are the `_BOOT_` PIDs. They are not the same as `RDM_PID_SOFTWARE_VERSION_LABEL`. I figure the default behavior would be for these PIDs to return the version of ESP-IDF that the ESP32 is running.
- `RDM_PID_BOOT_POWER_CYCLES` agreed. Like the `_BOOT_` parameters, this is simply a PID that would have a straight-forward default callback. Therefore, it would be easy for me implement. But perhaps this is adding features just for the sake of adding features.

Comment (someweisguy) [reply src/dmx/hal.c:464]: I'm resolving this thread because I'm happy with this code. I think we may move it later when we have a better idea of what it means to not have a DMX start address.

I've moved discussion of the default RDM parameters to #99 if you have any more thoughts!

**Review by someweisguy — APPROVED** [review]

**Comment by someweisguy** [comment]
Merged. Thank you for the great work!
